typeracer_bot.py

```diff
@@ -1,78 +1,81 @@
-        self.text = ""
-
-    def start(self):
-        # Open the link
-        self.driver.get(self.link)
-
-    def find_text(self):
-        # Resetting text
-        self.text = ""
-
-        # Get the HTML of the current page
-        src = self.driver.page_source
-        soup = BeautifulSoup(src, "html.parser")
-        span = soup.findAll("span")
-
-        for i in span:
-            if "unselectable" in str(i):
-                self.text += i.text
-
-        if self.text:
-            print("- valid text successfully found")
-
-    def keypress_type(self, delay):
-        for i in self.text:
-            time.sleep(delay)
-            self.keyboard.press(i)
-            self.keyboard.release(i)
-
-    def keysend_type(self, delay):
-        try:
-            # Click the inputbox when it can be clicked
-            wait = WebDriverWait(self.driver, 15)
-            elem = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "txtInput")))
-        except TimeoutException:
-            print("! inputbox not found within 15s")
-        else:
-            for i in self.text:
-                time.sleep(delay)
-                elem.send_keys(i)
-
-
-def main():
-    # Default link and delay time
-    link = "https://play.typeracer.com"
-    delay = 0.1
-
-    bot = TyperacerBot(link)
-    bot.start()
+def find_text(driver):
+    """Return typeracer text."""
+    # Get the HTML of the current page
+    src = driver.page_source
+    soup = BeautifulSoup(src, "html.parser")
+    span = soup.findAll("span")
+
+    text = ""
+    for i in span:
+        if "unselectable" in str(i):
+            text += i.text
+    return text
+
+def press_text(keyboard, text, delay):
+    """Type `text` using `keyboard` to simulate real keypresses."""
+    for i in text:
+        time.sleep(delay)
+        keyboard.press(i)
+        keyboard.release(i)
+
+def send_text(driver, text, delay):
+    """Send `text` to text input box first working driver in `drivers` dict."""
+    # Click the inputbox when it can be clicked
+    elem = WebDriverWait(driver, 15).until(EC.element_to_be_clickable((By.CLASS_NAME, "txtInput")))
+    for i in text:
+        time.sleep(delay)
+        elem.send_keys(i)
+
+# dict[name: lambda -> WebDriver]
+DRIVER_FUNCS = dict(
+    chrome=lambda: Chrome(executable_path="driver/chromedriver"),
+    firefox=lambda: Firefox(executable_path="driver/geckodriver"),
+    safari=lambda: Safari(executable_path="driver/safaridriver"),
+)
+
+def get_driver(preferred=None, drivers=DRIVER_FUNCS):
+    """Return first working driver in `drivers` dict."""
+    if preferred is not None:
+        with suppress(WebDriverException):
+            return drivers[preferred]()
+    for name, func in drivers.items():
+        with suppress(WebDriverException):
+            return func()
+    raise LookupError("Driver not found: {preferred=!r}")
+
+def main(link="https://play.typeracer.com", delay=0.1, driver=None):
+    if not isinstance(driver, WebDriver):
```

Comment:
can't u just do `if not driver`

Reply:
`driver` can have 3 possible types: `None`, `str`, or `WebDriver`.

When it's `None`, it'll go through the drivers like normal (checking if it works). `str` also sends the preferred driver to `get_driver`, meaning it'll attempt that driver first, then continue just like `None`. When a `WebDriver` is passed, it uses it directly, skipping the `get_driver`.

The code needs to check if the driver is a WebDriver (to skip or not to skip), so using `not isinstance(driver, WebDriver)` would work. If the check used `not driver`, a string would be used as the driver (which doesn't work).